docs/proposals/volumes.md

```diff
@@ -0,0 +1,373 @@
+Kubernetes.  For simplicity's sake, we will assume that if a volume requires ownership management,
+that it also requires relabeling and vice-versa.
+
+#### API requirements
+
+From the above, we know that ownership management must be applied:
+
+1.  To some volume types always
+2.  To some volume types never
+3.  To some volume types *sometimes*
+
+For volume types where the need for ownership/label management is not at the discretion of the user,
+there should be no need (or way) to specify whether ownership management is required in the API.
+Conversely there must be a way to specify this for volume types where it depends on the user's
+needs.
+
+### Determining correct ownership
+
+There are two components of ownership management:
+
+1.  When does Kubernetes need to manage the ownership of a volume?
+2.  How does Kubernetes determine the ownership of a volume?
+
+#### When to determine ownership
+
+Whether Kubernetes should manage the ownership of a volume for a distributed filesystem depends upon
+both the file system type and the cluster operator's policy.  For example:
+
+1.  Some organizations will manage ownership of volumes externally to the cluster
+2.  It is not possible to securely `chown` or `chmod` paths within some distributed filesystems
+
+#### How to determine ownership
+
+The Kubelet must analyze the pod spec to determine which UIDs need to use which volumes.  If a
+container's security context's `RunAsUser` field is not set, the Kubelet must inspect the image via
+the container runtime to determine which UID the image will run as.  Once the list of UIDs that need
+to use a volume is known, the kubelet can determine which ownership and permissions should be used
+to make the volume functional.
+
+If a volume is used only by a single UID within the pod, the ownership can be set to that UID.
+Otherwise, the volume should be owned by a group and the containers run in that group.
+
+If a non-numeric user is specified by an image, the behavior of container runtimes is to look up the
+UID from the container's `/etc/passwd` file.  It is not feasible for the kubelet to make this
+determination; we may not be able to correctly support non-numeric users in image metadata.  If a
+cluster operator allows users to create pods with images that specify non-numeric users, the system
+must use supplemental groups to make volumes shareable.
+
+### Setting ownership and permissions on volumes
+
+For volumes that are created by the system, it is sufficient to perform `chown` and `chmod`
+operations on the host's filesystem.  For some distributed file system volumes, such as ones based
+on remote block devices, `chown` and `chmod` applied locally will also work correctly.  File systems
+based on RPC, however, may not support client-side `chown` and `chmod`.  The system would need to
+schedule the `chown` and `chmod` operations to happen on the server side for these volumes.
```

Comment:
Do we want the server side permissions to be 777?  What are the security
implications?

On Aug 27, 2015, at 5:36 PM, Sami Wagiaalla <notifications@github.com>
wrote:

In docs/proposals/volumes.md
<https://github.com/kubernetes/kubernetes/pull/12944#discussion_r38142952>:

> +If a volume is used only by a single UID within the pod, the ownership can be set to that UID.
> +Otherwise, the volume should be owned by a group and the containers run in that group.
> +
> +If a non-numeric user is specified by an image, the behavior of container runtimes is to look up the
> +UID from the container's `/etc/passwd` file.  It is not feasible for the kubelet to make this
> +determination; we may not be able to correctly support non-numeric users in image metadata.  If a
> +cluster operator allows users to create pods with images that specify non-numeric users, the system
> +must use supplemental groups to make volumes shareable.
> +
> +### Setting ownership and permissions on volumes
> +
> +For volumes that are created by the system, it is sufficient to perform `chown` and `chmod`
> +operations on the host's filesystem.  For some distributed file system volumes, such as ones based
> +on remote block devices, `chown` and `chmod` applied locally will also work correctly.  File systems
> +based on RPC, however, may not support client-side `chown` and `chmod`.  The system would need to
> +schedule the `chown` and `chmod` operations to happen on the server side for these volumes.

Hmm.. this might be a challenge. I don't think we want kubernetes to get
involved with storage server. Maybe the server side permissions can be set
to 777 and mount can be 770 and owned by the pod group. Not sure if that is
possible. I'll look into doing this with NFS.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12944/files#r38142952>.